internal/controllers/lvmcluster/controller.go

```diff
@@ -337,6 +345,84 @@ func (r *Reconciler) setRunningPodImage(ctx context.Context) error {
 	return nil
 }
 
+func (r *Reconciler) checkForDeletedDeviceClasses(ctx context.Context, instance *lvmv1alpha1.LVMCluster) error {
+	logger := log.FromContext(ctx).WithValues("device class deletion for", instance.GetName())
+
+	currentVGs := &lvmv1alpha1.LVMVolumeGroupList{}
+	if err := r.List(ctx, currentVGs, client.InNamespace(r.GetNamespace())); err != nil {
+		return fmt.Errorf("failed to list existing LVMVolumeGroups: %w", err)
+	}
+
+	desiredVgNames := make(map[string]struct{})
+	for _, dc := range instance.Spec.Storage.DeviceClasses {
+		desiredVgNames[dc.Name] = struct{}{}
+	}
+
+	for _, currentVG := range currentVGs.Items {
+		if !metav1.IsControlledBy(&currentVG, instance) {
+			continue
+		}
+
+		if !currentVG.GetDeletionTimestamp().IsZero() {
+			continue
+		}
+
+		if _, ok := desiredVgNames[currentVG.Name]; ok {
+			continue
+		}
+
+		logger.Info("deleting lvm volume group and storage class because it no longer exists in device class list", "name", currentVG.Name)
+
+		err := r.Delete(ctx, &currentVG)
+		if err != nil {
+			return fmt.Errorf("failed to delete orphaned LVMVolumeGroup %s: %w", currentVG.Name, err)
```

Comment:
Returning early when unable to delete orphans might prevent storageclasses or volume snapshot classes without cleaning up, should we keep going and add it to an error list?

Reply:
That's a good point. I think we need to have OwnerReferences on objects so we can delete only one main object rather than deleting them separately